### PR TITLE
CPP: Support crement operations in CWE-190

### PR DESCRIPTION
**Pull request by geoffw0**
Support crement operations in the CWE-190 queries.  This fixes one of the 'regressions' we had when we changed from using Samate Juliet Test Suite version 1.2 to 1.3.

There's also a small improvement to the logic in `ArithmeticWithExtremeValues.ql` so that it understands the direction of the potential over/under-flow compared to the extreme value.

### CODE REVIEW

File: change-notes/1.18/analysis-cpp.md [hunk 1]
```diff
@@ -26,7 +26,11 @@
 | [Variable used in its own initializer] | Fewer false positive results | Results where a macro is used to indicate deliberate uninitialization are now excluded |
 | [Assignment where comparison was intended] | Fewer false positive results | Results are no longer reported if the variable is not yet defined. |
 | [Comparison where assignment was intended] | More correct results | "This query now includes results where an overloaded `operator==` is used in the wrong context. |
-
+| [User-controlled data in arithmetic expression] | More correct results | Crement operations are now understood as arithmetic operations in this query. |
+| [Uncontrolled data in arithmetic expression] | More correct results | Crement operations are now understood as arithmetic operations in this query. |
+| [Use of extreme values in arithmetic expression] | More correct results | Crement operations are now understood as arithmetic operations in this query. |
```
Comment (jbj) [inline change-notes/1.18/analysis-cpp.md]: I don't think "Crement" is a standard enough term to be used in a change note. Better write it out as "increment/decrement".

File: cpp/ql/src/semmle/code/cpp/security/Overflow.qll [hunk 2]
```diff
@@ -58,24 +58,26 @@ VariableAccess varUse(LocalScopeVariable v) {
 }
 
 /** is e not guarded against overflow by use? */
-predicate missingGuardAgainstOverflow(BinaryArithmeticOperation e, VariableAccess use) {
+predicate missingGuardAgainstOverflow(Operation e, VariableAccess use) {
   use = e.getAnOperand() and
   exists(LocalScopeVariable v | use.getTarget() = v |
     // overflow possible if large
     (e instanceof AddExpr and not guardedLesser(e, varUse(v))) or
+    (e instanceof IncrementOperation and not guardedLesser(e, varUse(v)) and v.getType().getUnspecifiedType() instanceof IntegralType) or
```
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/security/Overflow.qll]: If we're supporting `x + 1` and `x++` then why not `x += 1`? Same for subtraction.

File: cpp/ql/src/Security/CWE/CWE-190/ArithmeticWithExtremeValues.ql [hunk 3]
```diff
@@ -35,18 +35,18 @@ predicate isMinValue(MacroInvocationExpr mie) {
 
 class SecurityOptionsArith extends SecurityOptions {
   override predicate isUserInput(Expr expr, string cause) {
-    (isMaxValue(expr) and cause = "max value") or
-    (isMinValue(expr) and cause = "min value")
+    (isMaxValue(expr) and cause = "overflow") or
+    (isMinValue(expr) and cause = "underflow")
```
Comment (jbj) [inline cpp/ql/src/Security/CWE/CWE-190/ArithmeticWithExtremeValues.ql]: I think the old strings match better what this predicate actually computes. Unless there's a good reason to change these strings, I suggest keeping them as they are and having both a `string cause` and a `string effect` in the query, matching them up in the `or` formulas. It looks strange now that cause and effect are conflated.

File: change-notes/1.18/analysis-cpp.md (hunk 1 above)
Comment (geoffw0) [reply change-notes/1.18/analysis-cpp.md]: Done.

File: cpp/ql/src/semmle/code/cpp/security/Overflow.qll (hunk 2 above)
Comment (geoffw0) [reply cpp/ql/src/semmle/code/cpp/security/Overflow.qll]: Done.  FWIW I think some of these cases are naive, the right-hand-side of subtraction is ignored and most cases assume the other side of a binary arithmetic expression is (at least sometimes) positive.  I'll make a JIRA ticket to look into improving this.

File: cpp/ql/src/Security/CWE/CWE-190/ArithmeticWithExtremeValues.ql (hunk 3 above)
Comment (geoffw0) [reply cpp/ql/src/Security/CWE/CWE-190/ArithmeticWithExtremeValues.ql]: Done.

**Pull request merged**